src/Lecture4.hs

```diff
@@ -135,7 +140,47 @@ errors. We will simply return an optional result here.
 -}
 
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow [] = Nothing
+parseRow r =
+   let
+      w = wordsWhen (==',') r
+   in validateRow w
+
+wordsWhen     :: (Char -> Bool) -> String -> [String]
+wordsWhen p s =  case dropWhile p s of
+                      "" -> []
+                      s' -> w : wordsWhen p s''
+                            where (w, s'') = break p s'
+
+validateRow :: [String] -> Maybe Row
+validateRow l
+   | null  l = Nothing
+   | length l /= 3 = Nothing
+   | isNothing (readMaybe (l !! 1) :: Maybe TradeType) = Nothing
+   | Prelude.head l == "" = Nothing
+   | isJust (findString "." (l !! 2)) || isJust (findString "-" (l !! 2)) = Nothing
+   | isNothing (readMaybe (l !! 2) :: Maybe Int) = Nothing 
+   | otherwise = 
+      Just Row{
+         rowProduct   = Prelude.head l,
+         rowTradeType = readTradeType (l !! 1),
+         rowCost      = readCost (l !! 2)
+            }
```

Comment:
You have a nice implementation with different handling 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow r = do
    [rawProd, rawTrade, rawCost] <- Just (wordsWhen (==',') r)
    guard $ rawProd /= ""
    trade <- readMaybe rawTrade
    cost <- readMaybe rawCost
    guard $ cost >= 0
    pure $ Row rawProd trade cost
```